### ISSUE TITLE
ConcurrentModificationException when calling updateCustomerData

### ISSUE THREAD

**Issue by zilinx** [muxinc/mux-stats-sdk-exoplayer#152]
Mux library version: `2.5.2` for Exoplayer `2.14.1+`

There have been some reports of a crash when attempting to update the video data. The application will crash with an error:
```
Fatal Exception: java.util.ConcurrentModificationException
       at java.util.LinkedHashMap$LinkedHashIterator.nextNode(LinkedHashMap.java:757)
       at java.util.LinkedHashMap$LinkedKeyIterator.next(LinkedHashMap.java:780)
       at com.mux.stats.sdk.os.MuxDictionary.c(MuxDictionary.java:17)
       at com.mux.stats.sdk.core.model.BaseQueryData.keys(BaseQueryData.java:2)
       at com.mux.stats.sdk.core.model.BaseQueryData.d(BaseQueryData.java:2)
       at com.mux.stats.sdk.core.model.CustomerData.setCustomerVideoData(CustomerData.java:4)
       at com.mux.stats.sdk.muxstats.MuxStats.updateCustomerData(MuxStats.java:4)
       at com.mux.stats.sdk.muxstats.MuxStats.updateCustomerData(MuxStats.java:4)
       at com.mux.stats.sdk.muxstats.MuxBaseExoPlayer.updateCustomerData(MuxBaseExoPlayer.java:346)
```

This happens when calling:
```
val videoData = muxStats.customerData.customerVideoData
videoData.videoTitle = "newTitle"
muxStats.updateCustomerData(muxStats.customerData.customerPlayerData, videoData)
```

I realise this version of `updateCustomerData` call  has recently been deprecated. We will probably switch to use the new approach, but it'd be good to fix this problem, or at least verify it won't happen with `updateCustomerData(CustomerData data)`.

Note that it probably won't be possible to switch to the new call as long as this issue isn't fixed: https://github.com/muxinc/mux-stats-sdk-exoplayer/issues/137

**Comment by jsanford8** [comment]
Hi @zilinx - thanks or submitting this, and we'll take a look

**Comment by tomkordic** [comment]
@zilinx is the code snippet you provided enough to reproduce the crash ?
For ConcurrentModificationException we need at least two threads modifying same object, I do not see the other thread.

**Comment by zilinx** [comment]
I haven't reproduced the issue myself, it happens very infrequently. I only saw it in the crash reports from a release build. 
The snippet is there to help to understand how our code interacts with Mux SDK at the point where Mux-related crash stacktrace (posted earlier) starts.

**Comment by tomkordic** [comment]
This is really hard to reproduce, @zilinx can you explain a little bit a thread model of your app, where do you suspect it could be modifying the CustomerData other then what you shared ?

**Comment by zilinx** [comment]
I didn't imagine this would be possible to reproduce in a reliable way. Like I said, it doesn't happen very often.
As for our implementaiton. there are a few events (received as kotlin StateFlow) which will trigger an update. The events are handled off the main thread. Probably the thread it is running from is modifying the dictionary at the same time another function is trying to access it. Are Mux dictionary updates always issued from the same thread? If the dictionary is central to receiving/processing tracking updates, perhaps it should use an implementation similar to [ConcurrentHashMap](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ConcurrentHashMap.html), modified to work for a LinkedHashMap.

Alternatively I can issue the update from the main thread, but that would not fix the vulnerability in the SDK.

**Comment by daytime-em** [comment]
This issue has been fixed in [v2.5.7](https://github.com/muxinc/mux-stats-sdk-exoplayer/releases/tag/v2.5.7), which is now available on Artifactory, or the [release page](https://github.com/muxinc/mux-stats-sdk-exoplayer/releases/tag/v2.5.7).

Thank you for your report!
